### PR TITLE
chore: refactor create provider page (#1337)

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

This PR follows #1582 and refactor the `create new...` page

### Screenshot/screencast of this PR

creation-process

### What issues does this PR fix or reference?

it is part of #1337 

### How to test this PR?

1. settings -> resources -> create new podman machine -> play with it

### CODE REVIEW

File: packages/main/src/plugin/index.ts
Comment (benoitf) [inline packages/main/src/plugin/index.ts]: is it time consuming ?

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte]: Disk for podman are virtual, so for example I can create 25 machines on my computer without having 25*100GB available on my computer.

it means I can create a virtual disk of 100GB even if I have less than 20GB of available space

 How will it work ?

Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte]: Currently i only show the free disk available when you try to create apodman machine, so if you have 20GB it will only allow you to allocate 20GB.

But as it calculate the actual free disk space, if you create 25 machine without starting them you will always be able to allocate those 20GB every time you create a new machine.

I see the problem you're raising but i don't know how to behave tbh. Any idea?

File: packages/main/src/plugin/index.ts
Comment (lstocchi) [reply packages/main/src/plugin/index.ts]: in my case not much but i am on a new laptop, maybe this could be a factor ...  btw this can be changed based on what we decide to do on your other question.

**Comment by lstocchi** [comment]
This should be done and you can start playing with it. I only have to write some tests. 

@mairin could you please give it a look? Idk what we should display when the `create` button is clicked. I am a big fan of the linear progress bar so i added one here 😆 If there is some mockup i missed just let me know. Tried to look at it but i found nothing. Thanks!!

**Comment by mairin** [comment]
@lstocchi Wow this is looking really great! I just tested on macOS and it really feels nice. I did run into one odd condition though - I started it already having a Podman machine (I had to use brew to set it up bc the version we' re installing on 64 bit intel Macs right now is broken, its a known qemu bug waiting on a new release.) So I clicked it and it lets me create another Podman machine, but I accidentally created it with the same name as the first machine, and there' s no way to go back and correct the name of the machine, it just gets stuck like this:

Screen Shot 2023-03-21 at 7 22 14 PM

So I think the machine name field should check for any pre-existing machine names and then check against what the users input is for the machine name and block them from proceeding if there' s a name clash?

Anyway I am going to remove my machine and try again and give you more feedback on the progress bit of it!

**Comment by mairin** [comment]
@lstocchi I found the original mockup for the progress view, take a look (I forgot to move it to the new page for the design, it was left in an older scratch page, my bad!) 

Resources _ creating podman machine (1)

So the idea here was to grey out the form and have a progress bar more or less where you placed the one you put there. I actually like the design of yours, the way you have the progress going is a good solution if it' s not possible to provide time estimates or % completion. 

I don't know about showing the log live by default though... maybe could the log be something that could be expanded but by default is hidden? (Like if you were going to do like in the mockup, have something like a "> View log" link under the progess bar left-aligned, and that would expand to show the log?

(also there's a mini progress bar in the task bar at the bottom - you can disregard that, that was an earlier idea for the taskbar that we scrapped.)

**Comment by mairin** [comment]
Just a couple more things I discovered poking around on macOS, in case they are helpful:

- I deleted my original Podman VM so I had no Podman VM. I'm not sure why (some kind of sync or something must be needed?) but I got the same "error: podman-machine-default already exists" error message trying to create the new Podman VM even though I deleted the old one and had no actual Podman vm anyway. When I closed Podman Desktop out and re-opened it, this problem went away.

- *However* when I reopened Podman Desktop, the dashboard prompted me to create a new Podman VM and not thinking I couldn't t resist and created one from there. But it turned out ok. I went to Settings > Resources and made a 2nd new Podman VM using your create button, it worked fine, I remembered to use a different name for the VM and it all worked out ok.

- Another thing - when I went to start the first VM (the one on the left), it was a bit slow and I wasn't t sure what was going on... I pressed play but didn't t know if anything was happening:
  
Screen Shot 2023-03-21 at 8 44 01 PM

So I tried pressing play on the right one. Nothing happened, but I had the term open on the side and I saw the error there:

Screen Shot 2023-03-21 at 8 45 39 PM

So I think I couple of things we might need:

- When the VM is starting, it can take a while. The mockups have a little status area where it says "O Running" - 
Screen Shot 2023-03-21 at 8 46 48 PM - could we have a message there when it' s starting, maybe something like "O Starting" - maybe make it purple and pulse it?

- When one VM is running (or starting) the other can't be started. So we should let the user know that... some ideas:
  - Grey out the controls of any other VMs for a Podman provider if the first one is running. have a small message at the bottom of the card (or perhaps on hover over the controls?) that says "Stop the running provider before starting another one."
   - Assume if the user is starting one, they want to stop the other, so don' t grey out the controls, but when the user presses start on the 2nd one, give them a "Starting this provider requires the running provider to be stopped, would you like to continue?" and if they consent then stop the running provider and start the new one.

(Maybe @themr0c would have better phrasings for these?)

**Comment by lstocchi** [comment]
Thanks @mairin!! 

I was sure there was a mockup with the progress view, thanks for pointing me to it. I'll work on it 👍 
By having the cancel button there I should also solve your first problem. 

Regarding the start buttons, i'm trying to split this huge refactoring into small chunks. The next PR would be creating a loader button like the one you suggested https://codepen.io/mairindubh/pen/NWLdYvo and give better feedbacks to users.

File: packages/main/src/plugin/index.ts
Comment (benoitf) [reply packages/main/src/plugin/index.ts]: I was just wondering if we had to cache the value

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte
Comment (benoitf) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte]: I think it's not possible for my usecase so let's ignore it :-)

**Comment by benoitf** [comment]
I'm wondering for the sliders if there is a way to enter the value manually (like if I want to enter 5) and if the scrolling is pointing to 5.1 or 4.8

more a question for @mairin 

image

here on this mockup on the right I'm able to enter my value "manually"

**Comment by lstocchi** [comment]
@mairin updated it. It took a while as i needed to implement the cancel logic, sorry.
WDYT?

Not sure if i have to do something when the user click `Cancel`. I just close the "creation bar above" and allow to recreate a new one immediately. 

creation-process

P.S: i noticed i have to update the title to "Creating Podman Machine ..."

**Comment by mairin** [comment]
@benoitf i think targeting the field on the right is more difficult....  what if it was typeahead so if the slider has focus, you can just type the number to make the slider go to it?

**Comment by mairin** [comment]
@lstocchi wow, that looks amazing! the way you have the logs open/close looks great! big +1 from me!

File: packages/renderer/src/lib/ui/LinearProgress.svelte
Comment (benoitf) [inline packages/renderer/src/lib/ui/LinearProgress.svelte:1]: I'm not sure of the styles and why it's not using tailwind

I mean, all of these
`
    appearance: none;
    border: none;
    height: 0.10em;
    color: var(--pf-global--primary-color--100);
    font-size: 16px;
`
can use tailwind properties

File: packages/renderer/src/lib/ui/LinearProgress.svelte
Comment (benoitf) [inline packages/renderer/src/lib/ui/LinearProgress.svelte]: note sure about the usage of the mozilla properties, webkit and other properties
AFAIK in electron we have only one browser so we don't need to handle all browsers

File: packages/renderer/src/lib/preferences/PreferencesRenderingItemFormat.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesRenderingItemFormat.svelte]: shouldn't we check that there is a minimum as well ?

File: extensions/podman/src/podman-cli.ts
Comment (benoitf) [inline extensions/podman/src/podman-cli.ts]: is it a reject or resolve ?

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte]: should this svg be moved to a separate location in a separate component ?

File: extensions/podman/src/podman-cli.ts
Comment (lstocchi) [reply extensions/podman/src/podman-cli.ts]: not 100% but i thought that if the user is cancelling it voluntarily it's not an error, you want to resolve it without doing anything. If i reject the user will see the error message and i would need to filter it in the UI but still ... this is not an error.

**Comment by lstocchi** [comment]
@benoitf i should have fixed all your concerns. It's only missing some tests. Working on them.

File: extensions/podman/src/podman-cli.ts
Comment (benoitf) [reply extensions/podman/src/podman-cli.ts]: I suppose I expect an error because the method is shared by a lot of code and client needs to know what happened as it did not finished successfully
but error should be like a CancellationRequestError so it's filtered out by client

Comment (benoitf) [reply extensions/podman/src/podman-cli.ts]: Else, what we could do is always return a complex object with helpers like `isCancelled`

so instead of returning a string, we return an object that says if it was cancelled, succeed, error, etc.

 it could be an improvement done with https://github.com/containers/podman-desktop/issues/1585

**Comment by benoitf** [comment]
FYI https://github.com/containers/podman-desktop/issues/1877

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: probably you should replace the onMount and restore it in a beforeEach/afterEach call

you could replace the renderer method as well


with a custom `renderer` method doing
```
const originalOnMount = svelte.onMount;

beforeEach(() => {
svelte.onMount = {
try {
     originalOnMount();
     } finally {
     dispatch('pageLoaded');
     }
}
);
```

```
afterEach(() => {
svelte.onMount = originalOnMount

```

and for renderer
const originalRender = render;

```
beforeEach(() => {
render = {
  const result = originalRender(.....);
  return new Promise(resolve => {
    component.$on('pageLoaded', function() {
      resolve(result);
    });
  })  
```


in your code you're not handling error so you should add the dispatch inside a finally block of a try/finally

Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: Ops, i didn't notice i pushed my testing... my fault i was rebasing.
BTW this won't work or atleast i'm not able to make it work. You cannot use the dispatched there apparently.
It returns an error `Function called outside component initialization`

Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: I also tried with an eventEmitter, no luck

Comment (benoitf) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: Oh ok

Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: As i'm wasting too much time on it i propose three ways to progress
1) commenting the two existing tests and opening a PR to work on the onMount issue later
2) using the hack of having the dispatch on the svelte file
3) using the hack to export the variable and set it true when running the tests

Of course all three require to spend some time on the main problem but it's really blocking me and after 8 days i think it's better to move on.

**Comment by lstocchi** [comment]
> FYI #1877

I'm going to work on that so we can merge that before and rebase this.

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts
Comment (benoitf) [reply packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.spec.ts]: I would say option 1 or option 3 so svelte files are not too much 'hacked'

**Comment by lstocchi** [comment]
Ready to be reviewed. Thanks!!

**Review by benoitf — APPROVED** [review]
- need to fix formatting issue on files

I noticed I'm not able to remove the value for a 'file' type

usually I can select the name of a file and hit 'delete' key but here it does nothing

https://user-images.githubusercontent.com/436777/229436099-6f4fa6bb-d93a-4cf3-a02d-a824e153cbc4.mp4

so if I wrongly set a file I'm not able to remove the field value (I can change to another file but cannot remove)


Then (already commented) I find odd that if I select 2GB 100GB when creating the machine, once the machine is created it says 107.37GB and 2.15GB


When I create a machine, I have a top right 'Cancel' button and a bottom left 'Cancel' button.
As a user I may not understand why I have 2 cancel buttons and why I can't click cancel on the bottom left as well.

When creating a kind cluster, I tried the 'cancel' button and it does not work. It's continuing.

When I hit cancel I'm not able to see the logs of what I've cancelled and nothing is telling me that I've cancelled something. It goes in a state as if I'm on fresh create. Maybe it could be interesting to see : 'Action cancelled. See logs' and only if I click again on Create it disappear. (In case of Error we see the show Logs/close items)

I was able to cancel my podman creation of a machine.

Overall, lot of nice improvements

**Comment by benoitf** [comment]
And I forgot: When I have an error, the progress bar is still active while I think it shouldn't

https://user-images.githubusercontent.com/436777/229439301-1643e41e-e40d-4611-9dee-3860292cf110.mp4

File: packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesConnectionCreationRendering.svelte]: with https://github.com/containers/podman-desktop/pull/1916

you can now use a displayName or default to something else like 'instance' instead of Machine

**Comment by lstocchi** [comment]
I think i covered almost everything 

> Then (already commented) I find odd that if I select 2GB 100GB when creating the machine, once the machine is created it says 107.37GB and 2.15GB

Maybe we can round the values received (so you would see 107GB and 2 GB) or think at another solution later.

> When creating a kind cluster, I tried the 'cancel' button and it does not work. It's continuing.

My fault i didn't add the token to the kubernetes creation function. I'll create a new PR for it

> When I create a machine, I have a top right 'Cancel' button and a bottom left 'Cancel' button. As a user I may not understand why I have 2 cancel buttons and why I can't click cancel on the bottom left as well.

Not sure what to do about it. I renamed the button at the bottom as `Close` for the moment so to avoid confusion with cancel. We can also remove it and only keep the cross icon at the top.

For the rest i should have done everything
create-enhancement

**Comment by benoitf** [comment]
@lstocchi looks unit tests are failing

@mairin do you have something to share about the double "cancel" buttons on the same page (now Luca renamed one to close)

**Review by benoitf — APPROVED** [review]
approving (with follow-up PRs)

Some unit tests are not passing

**Comment by mairin** [comment]
> @mairin do you have something to share about the double "cancel" buttons on the same page (now Luca renamed one to close)

I thought about this a bit. I think the best way to handle this is to keep "cancel" where it is, but when it no longer applies grey it out (so the show logs control doesn't bounce around.) To do a close, I think I'd enclose the error message in a box (here I use the patternfly style one but you can just use a basic rounded corner one we use elsewhere in the UI and have the box include an X in the corner so it can be dismissed easily in a way that would be expected. I'd also move the error message up to be where the progress bar was, so the error is more closely tied with the task where the task was happening (the progress bar pulse.) 

Does that make sense? Here is what I was thinking:

Create Podman (1)

**Comment by deboer-tim** [comment]
I get distracted by how the inline error, creation progress, and show logs all 'move' the main UI up and down. e.g. I select the name field with the mouse, start typing, and then suddenly that field moves down below the cursor (and maybe back up) as I type.

Could the inline error be at the bottom, or inline with the machine name (if that's the only field it's associated with) so that it doesn't move other elements?

After you click Create could it hide all the fields, or maybe disable them and appear on top?

**Comment by lstocchi** [comment]
> I get distracted by how the inline error, creation progress, and show logs all 'move' the main UI up and down. e.g. I select the name field with the mouse, start typing, and then suddenly that field moves down below the cursor (and maybe back up) as I type.
> 
> Could the inline error be at the bottom, or inline with the machine name (if that's the only field it's associated with) so that it doesn't move other elements?
> 
> After you click Create could it hide all the fields, or maybe disable them and appear on top?

What do you mean by inline error? If it's the red one on the top it's not related to a single field. If the creation fails somehow it's where the error is displayed.

When you click on Create all fields are already disabled.

Regarding the position and hiding fields.
In some case it can be useful to see your inputs while the creation is running so if you notice you made a mistake you can cancel it, fix and restart. Or if it eventually fails you can start typing again without clicking any button to make the inputs visible again (if they were hidden before).

I mean, if you hide everything and you see the `Show Logs`, spinner, etc.. during the execution what happens if it fails? You click on close (hiding `Show logs`, spinner ....) and the inputs show up again? What if i want to see the logs at this point? The terminal would be there (moving the UI) like it is now?

Then if we don't hide the inputs i think it's better to have it on the top. It would be annoying to scroll everytime to the bottom

**Comment by deboer-tim** [comment]
This is the kind of 'moving UI' that I'd like to avoid (I mocked this up quickly, not real code):

https://user-images.githubusercontent.com/19958075/229599549-03d5c526-0409-4774-8d6c-951521f3a08f.mov

In a case like this the problem could be solved by leaving space for the error (i.e. use visibility, not hidden), moving it to the bottom, etc.

I thought the error message above was one of these, but watching the video again I see it currently isn't triggered until clicking Create. Shouldn't we be validating for this within the form, i.e. disable Create and show an error on the name field, like we do in preferences?

Looking at the rest again, @mairin shouldn't it just be using the Task Manager? This is another 'long running/background' task where it could launch and return you to the Resources page, maybe with some kind of 'task running' placeholder.

**Comment by mairin** [comment]
@deboer-tim Yeh ideally we'd validate the name before the form was submitted for this specific error with the vm name... and yeh ideally the long running task of installing the VM would be handled in the task manager, but having a "local" view of the progress on this page is helpful too. We don't have the full task manager UI design yet (just the tray popup implemented), so there isnt a place right now to get the full error logs / etc. for a task manager popup yet.

**Comment by benoitf** [comment]
@deboer-tim could be a next iteration ?
I think we're having benefits over the current state so probably it can be follow-up PR

**Comment by deboer-tim** [comment]
Yes, I am fine with that. I agree this is much better than before, so let's not try to improve further and merge, and maybe open two issues for later (one for form validation and one for task manager integration)?

**Comment by benoitf** [comment]
I think we're already integrated into the task manager

**Comment by lstocchi** [comment]
Merging this then so i can follow-up by adding the cancellation token support for kube connection creation process

**Pull request merged**
